lib/maths/common/CTools.cc

```diff
@@ -1393,63 +1399,88 @@ operator()(const lognormal& logNormal, double a, double b) const {
-    double erfb = logb == POS_INF ? 1.0 : std::erf((logb - c) / s);
+    double a_ = (loga - location) / s;
+    double b_ = (logb - location) / s;
+    double erfa = std::erf((loga - c) / s);
+    double erfb = std::erf((logb - c) / s);
 
     if (erfb - erfa < std::sqrt(EPSILON)) {
-        double expa = loga == NEG_INF ? 0.0 : std::exp(-a_ * a_);
-        double expb = logb == POS_INF ? 0.0 : std::exp(-b_ * b_);
+        double expa = std::exp(-a_ * a_);
+        double expb = std::exp(-b_ * b_);
         return expa == expb ? (2.0 * a / (a + b)) * b
                             : (expa + expb) / (expa / a + expb / b);
     }
 
-    double erfa_ = a_ == NEG_INF ? -1.0 : std::erf(a_);
-    double erfb_ = b_ == POS_INF ? 1.0 : std::erf(b_);
-    return mean * (erfb - erfa) / (erfb_ - erfa_);
+    double erfa_ = std::erf(a_);
+    double erfb_ = std::erf(b_);
+    return (erfb - erfa) == (erfb_ - erfa_) ? mean : mean * (erfb - erfa) / (erfb_ - erfa_);
 }
 
 double CTools::SIntervalExpectation::operator()(const gamma& gamma_, double a, double b) const {
     if (a > b) {
         std::swap(a, b);
     }
-    if (a == POS_INF) {
+    if (a >= POS_INF) {
         return POS_INF;
     }
     if (b <= 0.0) {
         return 0.0;
     }
+    if (a == b) {
+        return a;
+    }
 
     double shape = gamma_.shape();
-    double rate = 1.0 / gamma_.scale();
+    double scale = gamma_.scale();
+    if (a <= 1e-8) {
+        boost::math::gamma_distribution<> gammaShift_{shape + 1.0, scale};
+        return boost::math::mean(gamma_) * boost::math::cdf(gammaShift_, b) /
+               boost::math::cdf(gamma_, b);
+    }
+
+    double rate = 1.0 / scale;
     double mean = boost::math::mean(gamma_);
-    double gama = a <= 0.0 ? 0.0 : boost::math::gamma_p(shape + 1.0, rate * a);
-    double gamb = b == POS_INF ? 1.0 : boost::math::gamma_p(shape + 1.0, rate * b);
+    double gama = boost::math::gamma_p(shape + 1.0, rate * a);
+    double gamb = b >= POS_INF ? 1.0 : boost::math::gamma_p(shape + 1.0, rate * b);
 
     if (gamb - gama < std::sqrt(EPSILON)) {
-        double expa = a <= 0.0 ? 0.0 : std::exp((shape - 1.0) * std::log(a) - rate * a);
-        double expb = b == POS_INF ? 0.0 : std::exp((shape - 1.0) * std::log(b) - rate * b);
-        return (a * expa + b * expb) / (expa + expb);
+        double expa = std::exp((shape - 1.0) * std::log(a) - rate * a);
+        double expb = std::exp((shape - 1.0) * std::log(b) - rate * b);
+        return expa == expb ? (a + b) / 2.0 : (a * expa + b * expb) / (expa + expb);
     }
 
-    double gama_ = a <= 0.0 ? 0.0 : boost::math::gamma_p(shape, rate * a);
-    double gamb_ = b == POS_INF ? 1.0 : boost::math::gamma_p(shape, rate * b);
-    return mean * (gamb - gama) / (gamb_ - gama_);
+    double gama_ = boost::math::gamma_p(shape, rate * a);
+    double gamb_ = boost::math::gamma_p(shape, rate * b);
+    return (gamb - gama) == (gamb_ - gama_) ? mean : mean * (gamb - gama) / (gamb_ - gama_);
```

Comment:
Is it possible that `gamb_ - gama_ ` could be zero when `gamb - gama ` is not zero?

Reply:
See [cbaa29e](https://github.com/elastic/ml-cpp/pull/2286/commits/cbaa29e7862fec6da97f6fb5233edb3582b5c4ed).